apache/superset
# feat: new dataset/table/column models

### SUMMARY


This PR implements new models for SIP-68 (https://github.com/apache/superset/issues/14909):

- `Column`
- `Table`
- `Dataset`

The associated tables are created with a migration, and the existing datasets (`SqlaTable`, `SqlMetric`, `TableColumn`) are migrated to the new models:

- A physical dataset is mapped into a `Table` instance and a `Dataset` instance.
- A virtual dataset is mapped into a `Dataset` instance.
- Metrics and columns are mapped into `Column` instances (in the new model it represents physical columns, derived columns, and metrics).

The models are kept up-to-date via hooks added to `SqlaTable`, `SqlMetric`, and `TableColumn`. Every time these models are created, deleted, or updated, the new models are updated correspondingly. The sync is unidirectional, from the old models to the new ones.

The next step is to modify the backend to **read** from the new models. Initially the API will remain unmodified, but hopefully in the future we might be able to clean it up and modernize it (in some places we haven't migrated to the v1 API yet).

After that, the backend will be modified to **write** to the new models. Once the backend is reading and writing from the new models we can get rid of the old models (including the Druid models).

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF


No visual changes.

### TESTING INSTRUCTIONS


Reviewers can run `superset load-examples` and see the new models being populated:

```sql
mysql> SELECT `schema`, name FROM sl_tables;
+----------+------------------------+
| schema   | name                   |
+----------+------------------------+
| examples | wb_health_population   |
| examples | birth_names            |
| examples | long_lat               |
| examples | birth_france_by_region |
| examples | sf_population_polygons |
| examples | flights                |
| examples | bart_lines             |
| examples | channel_members        |
| examples | messages               |
| examples | channels               |
| examples | threads                |
| examples | unicode_test           |
| examples | users                  |
| examples | users_channels         |
| examples | cleaned_sales_data     |
+----------+------------------------+
15 rows in set (0.00 sec)

mysql> SELECT name, is_physical FROM sl_datasets;
+---------------------------+-------------+
| name                      | is_physical |
+---------------------------+-------------+
| wb_health_population      |           1 |
| birth_names               |           1 |
| long_lat                  |           1 |
| birth_france_by_region    |           1 |
| sf_population_polygons    |           1 |
| flights                   |           1 |
| bart_lines                |           1 |
| FCC 2018 Survey           |           0 |
| messages_channels         |           0 |
| channel_members           |           1 |
| covid_vaccines            |           0 |
| messages                  |           1 |
| channels                  |           1 |
| members_channels_2        |           0 |
| exported_stats            |           0 |
| threads                   |           1 |
| new_members_daily         |           0 |
| unicode_test              |           1 |
| users                     |           1 |
| video_game_sales          |           0 |
| users_channels            |           1 |
| cleaned_sales_data        |           1 |
| users_channels-uzooNNtSRO |           0 |
+---------------------------+-------------+
23 rows in set (0.00 sec)

mysql> SELECT expression FROM sl_columns WHERE is_aggregation;
+-----------------------+
| expression            |
+-----------------------+
| sum(`SP_POP_TOTL`)    |
| sum(`SH_DYN_AIDS`)    |
| sum(`SP_RUR_TOTL_ZS`) |
| sum(`SP_DYN_LE00_IN`) |
| sum(`SP_RUR_TOTL`)    |
| COUNT(*)              |
| COUNT(*)              |
| SUM(num)              |
| COUNT(*)              |
| AVG(`2004`)           |
| COUNT(*)              |
| COUNT(*)              |
| COUNT(*)              |
| COUNT(*)              |
| count(*)              |
+-----------------------+
15 rows in set (0.02 sec)
```

The PR has unit tests checking that the sync works on creation, deletion, and updates.

### ADDITIONAL INFORMATION


- [X] Has associated issue: https://github.com/apache/superset/issues/14909
